### PR TITLE
Initial DB Table

### PR DESCRIPTION
**Pull request by ralikio**
**Description**

Changes proposed in this pull request:

- Initial Schema for Kyma Bindings,
- go structs and logic.

**Related issue(s)**

#1198 #284

### CODE REVIEW

File: internal/storage/postsql/init.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ const (
 	SubaccountStatesTableName  = "subaccount_states"
 	CreatedAtField             = "created_at"
 	InstancesArchivedTableName = "instances_archived"
+	BindingssTableName         = "bindings"
```
Comment (jaroslaw-pieszka) [inline internal/storage/postsql/init.go:23]:
```suggestion
	BindingsTableName         = "bindings"
```

File: internal/storage/postsql/write.go [hunk 2]
```diff
@@ -21,6 +21,37 @@ type writeSession struct {
 	transaction *dbr.Tx
 }
 
+func (ws writeSession) DeleteBinding(ID string) dberr.Error {
+	_, err := ws.deleteFrom(BindingssTableName).
```
Comment (jaroslaw-pieszka) [inline internal/storage/postsql/write.go:25]:
```suggestion
	_, err := ws.deleteFrom(BindingsTableName).
```

File: internal/storage/postsql/write.go [hunk 3]
```diff
@@ -21,6 +21,37 @@ type writeSession struct {
...
+		Pair("version", binding.Version).
+		Exec()
+
+	if err != nil {
+		if err, ok := err.(*pq.Error); ok {
+			if err.Code == UniqueViolationErrorCode {
+				return dberr.AlreadyExists("binding with id %s already exist for runtime %s", binding.ID, binding.RuntimeID)
+			}
+		}
+		return dberr.Internal("Failed to insert record to Instance table: %s", err)
```
Comment (jaroslaw-pieszka) [inline internal/storage/postsql/write.go:49]:
```suggestion
		return dberr.Internal("Failed to insert record to Binding table: %s", err)
```

File: internal/model.go [hunk 4]
```diff
@@ -579,3 +579,18 @@ type DeletedStats struct {
...
+
+	CreatedAt time.Time
+	UpdatedAt time.Time
+	ExpiredAt *time.Time
+
+	Kubeconfig         string
+	ExpireationSeconds int64
+	GenerationMethod   string
+
+	Version int
```
Comment (jaroslaw-pieszka) [inline internal/model.go:595]: YAGNI?

File: internal/storage/driver/postsql/binding.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	return dbmodel.BindingDTO{
+		Kubeconfig: string(encrypted),
+		ID:         binding.ID,
+		RuntimeID:  binding.RuntimeID,
+		CreatedAt:  binding.CreatedAt,
+		UpdatedAt:  binding.UpdatedAt,
+		ExpiredAt:  binding.ExpiredAt,
+		Version:    binding.Version,
```
Comment (jaroslaw-pieszka) [inline internal/storage/driver/postsql/binding.go:100]: Not present in the DB structure yet.

File: internal/storage/postsql/read.go [hunk 6]
```diff
@@ -21,6 +21,36 @@ type readSession struct {
...
+		if err == dbr.ErrNotFound {
+			return dbmodel.BindingDTO{}, dberr.NotFound("Cannot find Binding for bindingId:'%s'", bindingID)
+		}
+		return dbmodel.BindingDTO{}, dberr.Internal("Failed to get Instance: %s", err)
+	}
+
+	return binding, nil
+}
+
+func (r readSession) ListBindings(runtimeID string) ([]dbmodel.BindingDTO, error) {
```
Comment (jaroslaw-pieszka) [inline internal/storage/postsql/read.go:43]: Access not by index (yet)

**Comment by piotrmiskiewicz** [comment]
consider adding Instance ID

File: internal/storage/postsql/read.go [hunk 7]
```diff
@@ -21,6 +21,39 @@ type readSession struct {
...
+	var bindings []dbmodel.BindingDTO
+	if len(runtimeID) == 0 {
+		return bindings, fmt.Errorf("runtimeID cannot be empty")
+	}
+	stmt := r.session.Select("*").From(BindingsTableName)
+	if len(runtimeID) != 0 {
+		stmt.Where(dbr.Eq("runtime_id", runtimeID))
```
Comment (piotrmiskiewicz) [inline internal/storage/postsql/read.go:50]: please review all runtime_id occurences

File: internal/storage/driver/postsql/binding.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	sess := s.NewWriteSession()
+	err = sess.InsertBinding(dto)
+	if err != nil {
+		return fmt.Errorf("while saving binding with ID %s: %w", binding.ID, err)
+	}
+
+	return nil
+}
+
+func (s *Binding) Delete(instanceID string) error {
```
Comment (piotrmiskiewicz) [inline internal/storage/driver/postsql/binding.go:64]: instance ID or binding ID?

File: internal/storage/driver/postsql/binding.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+	sess := s.NewWriteSession()
+	return sess.DeleteBinding(instanceID)
+}
+
+func (s *Binding) List(runtimeID string) ([]internal.Binding, error) {
```
Comment (piotrmiskiewicz) [inline internal/storage/driver/postsql/binding.go:69]: consider ListByInstanceID

File: internal/storage/driver/postsql/binding.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+	return nil
+}
+
+func (s *Binding) Delete(instanceID string) error {
+	sess := s.NewWriteSession()
+	return sess.DeleteBinding(instanceID)
+}
+
+func (s *Binding) List(instanceID string) ([]internal.Binding, error) {
```
Comment (piotrmiskiewicz) [inline internal/storage/driver/postsql/binding.go:69]: ListByInstanceID

**Review by piotrmiskiewicz — APPROVED** [review]

**Review by piotrmiskiewicz — DISMISSED** [review]

**Review by piotrmiskiewicz — DISMISSED** [review]

**Review by piotrmiskiewicz — APPROVED** [review]

**Review by piotrmiskiewicz — APPROVED** [review]

**Pull request merged**